src/Microsoft.DotNet.Interactive/Server/NotebookParserServer.cs

```diff
@@ -0,0 +1,409 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Microsoft.DotNet.Interactive.Documents;
+using Microsoft.DotNet.Interactive.Documents.Jupyter;
+
+namespace Microsoft.DotNet.Interactive.Server
+{
+    public class NotebookParserServer
+    {
+        public TextReader Input { get; }
+        public TextWriter Output { get; }
+
+        public static IReadOnlyCollection<KernelName> WellKnownKernelNames = new KernelName[]
+        {
+            new KernelName("csharp", new[] { "c#", "C#", "cs" }),
+            new KernelName("fsharp", new[] { "f#", "F#", "fs" }),
+            new KernelName("pwsh", new[] { "powershell" }),
+            new KernelName("javascript", new[] { "js" }),
+            new KernelName("html", Array.Empty<string>()),
+            new KernelName("sql", Array.Empty<string>()),
+            new KernelName("kql", Array.Empty<string>()),
+        };
+
+        public NotebookParserServer(TextReader input, TextWriter output)
+        {
+            Input = input ?? throw new ArgumentNullException(nameof(input));
+            Output = output ?? throw new ArgumentNullException(nameof(output));
+        }
+
+        public async Task RunAsync()
+        {
+            while (true)
+            {
+                var line = await Input.ReadLineAsync();
+                if (line is not null)
+                {
+                    NotebookParserServerResponse response;
+                    string requestId = null;
+                    try
+                    {
+                        var request = NotebookParseOrSerializeRequest.FromJson(line);
+                        requestId = request.Id;
+                        response = HandleRequest(request);
+                    }
+                    catch (Exception ex)
+                    {
+                        response = new NotebookErrorResponse(requestId, ex.ToString());
+                    }
+
+                    // if there was no id, we can't reasonably respond with anything
+                    if (requestId is not null)
+                    {
+                        response.Id = requestId;
+                        var responsePayload = response.ToJson();
+                        await Output.WriteLineAsync(responsePayload);
+                        await Output.FlushAsync();
+                    }
+                }
+            }
+        }
+
+        public static NotebookParserServerResponse HandleRequest(NotebookParseOrSerializeRequest request)
+        {
+            try
+            {
+                switch (request)
+                {
+                    case NotebookParseRequest parse:
+                        {
+                            var contentStream = new MemoryStream(parse.RawData);
+                            var document = request.Extension.ToLowerInvariant() switch
```

Comment:
Is there a way to specify the format and override the expectation based on file extension?

Reply:
The format is dictated by the `Extension` property, but it sounds like calling it `Format` with values of `dib` and `ipynb` would be more clear.